lib/sha256.c

```diff
@@ -37,9 +38,190 @@
 
 #endif
 
-#ifdef USE_OPENSSL_SHA256
+#if defined(USE_GNUTLS_NETTLE)
+
+#include <nettle/sha.h>
+
+#include "curl_memory.h"
+
+/* The last #include file should be: */
+#include "memdebug.h"
+
+typedef struct sha256_ctx SHA256_CTX;
+
+static void SHA256_Init(SHA256_CTX *ctx)
+{
+  sha256_init(ctx);
+}
+
+static void SHA256_Update(SHA256_CTX *ctx,
+                          const unsigned char *input,
+                          unsigned int inputLen)
+{
+  sha256_update(ctx, inputLen, input);
+}
+
+static void SHA256_Final(unsigned char digest[32], SHA256_CTX *ctx)
+{
+  sha256_digest(ctx, SHA256_DIGEST_SIZE, digest);
+}
+
+#elif defined(USE_GNUTLS)
+
+#include "curl_memory.h"
+
+/* The last #include file should be: */
+#include "memdebug.h"
+
+typedef struct gcry_md_hd_t SHA256_CTX;
+
+static int SHA256_Init(SHA256_CTX *ctx)
+{
+  gcry_md_open(ctx, GCRY_MD_SHA256, 0);
+}
+
+static void SHA256_Update(SHA256_CTX *ctx,
+                          const unsigned char *input,
+                          unsigned int inputLen)
+{
+  gcry_md_write(*ctx, input, inputLen);
+}
+
+static void SHA256_Final(unsigned char digest[32], SHA256_CTX *ctx)
+{
+  memcpy(digest, gcry_md_read(*ctx, 0), SHA256_DIGEST_LENGTH);
+  gcry_md_close(*ctx);
+}
+
+#elsif defined(USE_OPENSSL_SHA256)
+
 /* When OpenSSL is available we use the SHA256-function from OpenSSL */
 #include <openssl/sha.h>
+
+#elif defined(USE_SECTRANSP)
+
+#include <CommonCrypto/CommonDigest.h>
+
+#include "curl_memory.h"
+
+/* The last #include file should be: */
+#include "memdebug.h"
+
+typedef struct {
+  void *data;
+  unsigned long size;
+} SHA256_CTX;
+
+static void SHA256_Init(SHA256_CTX *ctx)
+{
+  ctx->data = NULL;
+  ctx->size = 0;
+}
+
+static void SHA256_Update(SHA256_CTX *ctx,
+                          const unsigned char *input,
+                          unsigned int inputLen)
+{
+  if(ctx->data == NULL) {
+    ctx->data = malloc(inputLen);
+    if(ctx->data != NULL) {
+      memcpy(ctx->data, input, inputLen);
+      ctx->size = size;
+    }
+  }
+}
+
+static void SHA256_Final(unsigned char digest[32], SHA256_CTX *ctx)
+{
+  if(ctx->data != NULL) {
+    (void) CC_SHA256(ctx->data, (CC_LONG) ctx->size, digest);
```

Comment:
You actually don't have to do this if you don't want to do this. CommonDigest.h also has contextual functions for SHA256, just like OpenSSL and GnuTLS. [Here's the header](https://opensource.apple.com/source/CommonCrypto/CommonCrypto-60165/include/CommonDigest.h.auto.html).

Reply:
Cheers Nick,

I wasn't sure which this the preferred way - as I noticed that the MD4 code does it this way, but the MD5 code does it the way you suggest.

Are they any version issues of during it using the compatible API?

Reply:
That API goes all the way back to macOS 10.4 15 years ago. No version issues.

Reply:
Cool - ta Nick,

Would you recommend tidying up the MD4 code as well so we have consistency over the different digests?

Reply:
I would. I actually tried to make a PR that would do this for the MD4 code, but I'm unable to push. I keep getting a permission denied error. Do I still have permission to make PRs for this project?

Reply:
You are a member of the curl organization and have write access to all the projects. Typically you could push a branch for the pr to your own fork and create the pr from that. What is the error?

Reply:
Never mind. I was having key trouble yesterday, but I cleared it up on my own. PR is up.